### PR TITLE
[BOLT][DWARF] Add support for DW_IDX_parent

### PR DESCRIPTION
**Pull request by ayermolo**
This adds support for DW_IDX_parent. If DIE has a parent then DW_IDX_parent in Entry will point to Entry for that parent DIE. Otherwise it will have DW_FORM_flag_present in abbrev. Which takes zero space in Entry.

This came from
https://discourse.llvm.org/t/rfc-improve-dwarf-5-debug-names-type-lookup-parsing-speed/74151

### CODE REVIEW

**Review by dcci — APPROVED** [review]

File: bolt/lib/Core/DebugNames.cpp [hunk 1]
```diff
@@ -464,13 +502,33 @@ void DWARF5AcceleratorTable::writeEntries() {
...
+        if (!ParentOffset)
+          continue;
+        if (const auto Iter = EntryRelativeOffsets.find(*ParentOffset);
+            Iter != EntryRelativeOffsets.end()) {
+          const uint64_t PatchOffset = Entry->getPatchOffset();
+          uint32_t *Ptr = reinterpret_cast<uint32_t *>(
+              &EntriesBuffer.get()->data()[PatchOffset]);
+          *Ptr = Iter->second;
+        } else {
+          BC.errs() << "Could not find entry with offset " << *ParentOffset
```
Comment (maksfb) [inline bolt/lib/Core/DebugNames.cpp:526]: Is it a warning or an error and when can it get triggered? If the former, change the message to: "BOLT-WARNING: could not find...".

File: bolt/include/bolt/Core/DIEBuilder.h [hunk 2]
```diff
@@ -209,7 +209,10 @@ class DIEBuilder {
   void updateReferences();
 
   /// Update the Offset and Size of DIE, populate DebugNames table.
```
Comment (maksfb) [inline bolt/include/bolt/Core/DIEBuilder.h:211]: Update the comment with the new parameter description.

File: bolt/include/bolt/Core/DebugNames.h [hunk 3]
```diff
@@ -51,8 +54,10 @@ class DWARF5AcceleratorTable {
   /// Add DWARF5 Accelerator table entry.
   /// Input is DWARFUnit being processed, DIE that belongs to it, and potential
   /// SkeletonCU if the Unit comes from a DWO section.
```
Comment (maksfb) [inline bolt/include/bolt/Core/DebugNames.h:56]: Needs update as well.

File: bolt/lib/Core/DebugNames.cpp [hunk 4]
```diff
@@ -456,6 +483,17 @@ void DWARF5AcceleratorTable::writeEntry(const BOLTDWARF5AccelTableData &Entry) {
                              llvm::endianness::little);
       break;
     }
+    case dwarf::DW_IDX_parent: {
+      assert(
+          (AttrEnc.Form == dwarf::DW_FORM_ref4 && Entry.getParentDieOffset()) ||
+          AttrEnc.Form == dwarf::DW_FORM_flag_present);
+      if (std::optional<uint64_t> ParentOffset = Entry.getParentDieOffset()) {
+        Entry.setPatchOffset(EntriesBuffer->size());
+        support::endian::write(*Entriestream, static_cast<uint32_t>(0xBADF00D),
```
Comment (maksfb) [inline bolt/lib/Core/DebugNames.cpp:492]: Is `0xBADF00D` really a part of DWARF spec?

Comment (ayermolo) [reply bolt/lib/Core/DebugNames.cpp:492]: It is not. Can set to UINT32_MAX?

File: bolt/lib/Core/DebugNames.cpp [hunk 5]
```diff
@@ -523,8 +523,9 @@ void DWARF5AcceleratorTable::writeEntries() {
               &EntriesBuffer.get()->data()[PatchOffset]);
           *Ptr = Iter->second;
         } else {
-          BC.errs() << "Could not find entry with offset " << *ParentOffset
-                    << "\n";
+          BC.errs() << "BOLT-WARNING: [internal-dwarf-warning]: Could not find "
```
Comment (maksfb) [inline bolt/lib/Core/DebugNames.cpp:526]: "Could" -> "could"

Under what conditions can the user get this warning? Bad input?

File: bolt/lib/Core/DebugNames.cpp (hunk 4 above)
Comment (maksfb) [reply bolt/lib/Core/DebugNames.cpp:492]: Sure if that's what DWARF spec recommends.

File: bolt/lib/Core/DebugNames.cpp (hunk 5 above)
Comment (ayermolo) [reply bolt/lib/Core/DebugNames.cpp:526]: I can't think of any. Unless there a bug.

**Review by maksfb — APPROVED** [review]
Thanks.

**Pull request merged**